api/src/test/java/com/epam/pipeline/acl/folder/FolderApiServiceTest.java

```diff
@@ -0,0 +1,632 @@
+    public void shouldDenyUpdateWhenPermissionIsNotGranted() {
+        initAclEntity(folder);
+
+        assertThrows(AccessDeniedException.class, () -> folderApiService.update(folder));
+    }
+
+    @Test
+    @WithMockUser(roles = ADMIN_ROLE)
+    public void shouldGetProjectForAdmin() {
+        doReturn(folderWithMetadata).when(mockFolderManager).getProject(ID, AclClass.FOLDER);
+
+        assertThat(folderApiService.getProject(ID, AclClass.FOLDER)).isEqualTo(folderWithMetadata);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldGetProjectWhenPermissionIsGranted() {
+        doReturn(folderWithMetadata).when(mockEntityManager).load(AclClass.FOLDER, ID);
+        doReturn(folderWithMetadata).when(mockFolderManager).getProject(ID, AclClass.FOLDER);
+        initAclEntity(folderWithMetadata, AclPermission.READ);
+        mockSecurityContext();
+
+        assertThat(folderApiService.getProject(ID, AclClass.FOLDER)).isEqualTo(folderWithMetadata);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldDenyGetProjectWhenPermissionIsNotGranted() {
+        doReturn(folderWithMetadata).when(mockEntityManager).load(AclClass.FOLDER, ID);
+        doReturn(folderWithMetadata).when(mockFolderManager).getProject(ID, AclClass.FOLDER);
+        initAclEntity(folderWithMetadata);
+        mockSecurityContext();
+
+        assertThrows(AccessDeniedException.class, () -> folderApiService.getProject(ID, AclClass.FOLDER));
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldGetProjectWithHierarchyWhenPermissionIsGranted() {
+        final FolderWithMetadata childFolderWithPermission = initFolderWithReadPermission();
+        final FolderWithMetadata childFolderWithoutPermission = initFolderWithoutReadPermission();
+        final List<Folder> folders = Arrays.asList(childFolderWithPermission,
+                childFolderWithoutPermission,
+                emptyChildFolderWithoutPermission);
+        final FolderWithMetadata parentFolder = initParentFolder(folders);
+        doReturn(parentFolder).when(mockEntityManager).load(AclClass.FOLDER, ID);
+        doReturn(parentFolder).when(mockFolderManager).getProject(ID, AclClass.FOLDER);
+        initAclEntity(emptyChildFolderWithoutPermission);
+        initAclEntity(parentFolder, AclPermission.READ);
+        initEntities();
+        mockSecurityContext();
+
+        final FolderWithMetadata returnedFolder = folderApiService.getProject(ID, AclClass.FOLDER);
+        final Map<Long, AbstractHierarchicalEntity> childrenById = returnedFolder.getChildren().stream()
+                .collect(Collectors.toMap(AbstractSecuredEntity::getId, Function.identity()));
+        final List<? extends AbstractSecuredEntity> withoutPermissionLeaves =
+                childrenById.get(childFolderWithoutPermission.getId()).getLeaves();
+
+        assertThat(returnedFolder).isEqualTo(parentFolder);
```

Comment:
Let's add permissions mask check

Reply:
Also, let's add check for children list size. (Just to make sure that nothing extra has come in to result)